File: 17-retrieve-http-status-codes/java/src/test/java/StatusCodes.java
import org.junit.Before;
import org.junit.After;
import org.junit.Test;
import org.openqa.selenium.Proxy;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;
import net.lightbody.bmp.BrowserMobProxy;
import net.lightbody.bmp.BrowserMobProxyServer;
import net.lightbody.bmp.client.ClientUtil;
import net.lightbody.bmp.core.har.Har;
import net.lightbody.bmp.proxy.CaptureType;
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class StatusCodes {

    WebDriver driver;
    BrowserMobProxy proxy;

    @Before
    public void setUp() throws Exception {
        // start the proxy immediately
        proxy = new BrowserMobProxyServer();
        proxy.start(0);

        // create a proxy instance and configure Selenium to use it
        Proxy seleniumProxy = ClientUtil.createSeleniumProxy(proxy);
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability(CapabilityType.PROXY, seleniumProxy);

        // create an instance of Selenium connected to the proxy
        driver = new FirefoxDriver(capabilities);

        // enable detailed HAR capture
        proxy.enableHarCaptureTypes(
                CaptureType.REQUEST_CONTENT,
                CaptureType.RESPONSE_CONTENT);
    }

    @After
    public void tearDown() throws Exception {
        proxy.stop();
        driver.quit();
    }

    @Test
    public void ResourceNotFound() {
        proxy.newHar();
        driver.navigate().to("http://the-internet.herokuapp.com/status_codes/404");
        Har har = proxy.getHar();
        int response = har.getLog().getEntries().get(0).getResponse().getStatus();
        assertThat(response, is(404));
    }

}

File: 17-retrieve-http-status-codes/kotlin/src/test/kotlin/StatusCodes.kt
import net.lightbody.bmp.BrowserMobProxyServer
import net.lightbody.bmp.client.ClientUtil
import net.lightbody.bmp.proxy.CaptureType
import org.junit.Before
import org.junit.After
import org.junit.Test
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.firefox.FirefoxOptions
import org.openqa.selenium.firefox.FirefoxProfile
import org.openqa.selenium.remote.CapabilityType
import kotlin.test.assertEquals

class StatusCodes {

    private val driver by lazy {
        // create an instance of Selenium connected to the proxy
        val profile = FirefoxProfile()
        // create a proxy instance and configure Selenium to use it
        val seleniumProxy = ClientUtil.createSeleniumProxy(proxy)
        val options = FirefoxOptions().setProfile(profile)
        options.setCapability(CapabilityType.PROXY, seleniumProxy)
        FirefoxDriver(options)
    }

    private val proxy by lazy {
        // start the proxy immediately
        BrowserMobProxyServer()
    }

    @Before
    @Throws(Exception::class)
    fun setUp() {

        proxy.start(0)

        // enable detailed HAR capture
        proxy.enableHarCaptureTypes(
            CaptureType.REQUEST_CONTENT,
            CaptureType.RESPONSE_CONTENT
        )
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        // Quit session (closes browser)
        proxy.stop()
        driver.quit()
    }

    @Test
    fun `Resource not found`() {
        proxy.newHar()
        driver.navigate().to("http://the-internet.herokuapp.com/status_codes/404")
        val har = proxy.har
        val response = har.log.entries[0].response.status
        assertEquals(
            404,
            response
        )
    }
}